collector/ipvs.go

```diff
@@ -0,0 +1,258 @@
+// +build !noipvs
+
+package collector
+
+import (
+	"bufio"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	procNetIpvs      = "/proc/net/ip_vs"
+	procNetIpvsStats = "/proc/net/ip_vs_stats"
+	ipvsSubsystem    = "ipvs"
+)
+
+var (
+	ipvsStatFieldNames = []string{
+		"connections_total",
+		"incoming_packets_total",
+		"outgoing_packets_total",
+		"incoming_bytes_total",
+		"outgoing_bytes_total",
+	}
+	ipvsBackendLabelNames = []string{
+		"local_address",
+		"local_port",
+		"remote_address",
+		"remote_port",
+		"proto",
+	}
+)
+
+type ipvsCollector struct {
+	config                                                           Config
+	backendConnectionsActive, backendConnectionsInAct, backendWeight *prometheus.GaugeVec
+	stats                                                            map[string]prometheus.Counter
+}
+
+type ipvsBackendStatus struct {
+	localAddress  net.IP
+	localPort     uint16
+	proto         string
+	remoteAddress net.IP
+	remotePort    uint16
+	activeConn    uint16
+	inActConn     uint16
+	weight        uint16
+}
+
+func init() {
+	Factories["ipvs"] = NewIpvsCollector
+}
+
+func NewIpvsCollector(config Config) (Collector, error) {
+	c := ipvsCollector{
+		config: config,
+		backendConnectionsActive: &prometheus.GaugeVec{},
+		backendConnectionsInAct:  &prometheus.GaugeVec{},
+		backendWeight:            &prometheus.GaugeVec{},
+		stats:                    map[string]prometheus.Counter{},
+	}
+
+	for _, metricName := range ipvsStatFieldNames {
+		c.stats[metricName] = prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Namespace: Namespace,
+				Subsystem: ipvsSubsystem,
+				Name:      metricName,
```

Comment:
(which would mean, you'd either have to list them all out separately, or you put both the metric name and the docstring into your list at the top (as a list of structs with two fields))